internal/app/singularity/registry.go

```diff
@@ -0,0 +1,146 @@
+func (l *Library) Pull(ctx context.Context, from, to string) error {
+	// strip leading "library://" and append default tag, as necessary
+	libraryPath := library.NormalizeLibraryRef(from)
+
+	// check if image exists in library
+	imageMeta, err := l.client.GetImage(ctx, libraryPath)
+	if err == scs.ErrNotFound {
+		return fmt.Errorf("image %s does not exist in the library", libraryPath)
+	}
+	if err != nil {
+		return fmt.Errorf("could not get image info: %v", err)
+	}
+
+	if l.cache == nil {
+		// don't use cached image
+		err := l.pullAndVerify(ctx, imageMeta, libraryPath, to)
+		if err != nil {
+			return fmt.Errorf("unable to download image: %v", err)
+		}
+	} else {
+		// check and use cached image
+		imageName := uri.GetName("library://" + libraryPath)
+		err := l.copyFromCache(imageMeta.Hash, imageName, to)
+		if err != nil {
+			if err != errNotInCache {
+				return fmt.Errorf("could not copy image from cache: %v", err)
+			}
+			imagePath := l.cache.LibraryImage(imageMeta.Hash, imageName)
+			err := l.pullAndVerify(ctx, imageMeta, libraryPath, imagePath)
+			if err != nil {
+				return fmt.Errorf("could not pull image: %v", err)
+			}
+			err = l.copyFromCache(imageMeta.Hash, imageName, to)
+			if err != nil {
+				return fmt.Errorf("could not copy image from cache: %v", err)
+			}
+		}
+	}
+
+	_, err = signing.IsSigned(to, l.keystoreURI, 0, false, l.client.AuthToken)
+	if err != nil {
+		sylog.Warningf("%v", err)
+		return ErrLibraryPullUnsigned
+	}
+
+	sylog.Infof("Download complete: %s\n", to)
+	return nil
+}
+
+// pullAndVerify downloads library image and verifies it by comparing checksum
+// in imgMeta with actual checksum of the downloaded file. The resulting image
+// will be saved to the location provided.
+func (l *Library) pullAndVerify(ctx context.Context, imgMeta *scs.Image, from, to string) error {
+	sylog.Infof("Downloading library image")
+	go interruptCleanup(to)
+
+	err := library.DownloadImage(ctx, l.client, to, from, printProgress)
+	if err != nil {
+		return fmt.Errorf("unable to download image: %v", err)
+	}
+
+	fileHash, err := scs.ImageHash(to)
+	if err != nil {
+		return fmt.Errorf("error getting image hash: %v", err)
+	}
+	if fileHash != imgMeta.Hash {
+		return fmt.Errorf("file hash(%s) and expected hash(%s) does not match", fileHash, imgMeta.Hash)
+	}
+	return nil
+}
+
+// copyFromCache checks whether an image with the given name and checksum exists in the cache.
+// If so, the image will be copied to the location provided.
+func (l *Library) copyFromCache(hash, name, to string) error {
+	exists, err := l.cache.LibraryImageExists(hash, name)
+	if err != nil {
+		return fmt.Errorf("unable to check if %s exists: %v", name, err)
+	}
+	if !exists {
+		return errNotInCache
+	}
+
+	// Perms are 777 *prior* to umask in order to allow image to be
+	// executed with its leading shebang like a script
+	dstFile, err := os.OpenFile(to, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
```

Comment:
In order to deal gracefully with various errors that might happen here, this should be creating a temporary file in the destination directory, and set things up to remove that temporary file when this function exits.

If the whole operation succeeds, the temporary file should be renamed to the final name (`to` here). In that way, if something fails, the temporary file gets removed. If nothing fails, you end up with the correct contents in the destination file.

In other words, the defer'ed function should be something like:

```
func () {
    if err := dstFile.Close(); err == nil {
        os.Rename(tmpname, to)
    }
    os.Remove(tmpname)
}
```

Close might result in error and Rename might result in error, but no matter what, the file gets removed.

By defer'ing this, the error on line 137 does not result in a truncated file, for example. Same with the error on line 143.